File: src/main/scala/com/MyLinked/LinkedList.scala
package com.MyLinked

import scala.reflect.ClassTag

class LinkedList[T] {

   private class Node(){
    var e:T = _
    var next:Node = _

    def this (t:T) {
      this ()    //调用主构造函数
      this.e=t
    }

    def this (e:T, node:Node) {
      this (e)    //调用辅助构造函数
      this.next=node
    }

    override def toString: String = {
      e.toString
    }

  }

  private var dummyHead = new Node()
  private var size = 0

  def getSize(): Int = {size}

  def isEmpty(): Boolean = {size==0}

  def add(index:Int,e:T): Unit ={
    if(index < 0 || index > size) throw new IllegalArgumentException("add failed,index need in [0,%d]".format(size))

    var prev = dummyHead
    for(i <- Range(0,index)){
        prev = prev.next
    }
    prev.next = new Node(e,prev.next)
    size+=1
  }

  def addFirst(e:T): Unit ={
    add(0,e)
  }

  def addLast(e:T): Unit ={
    add(size,e)
  }

  def get(index:Int):T = {
    if(index < 0 || index >= size) throw new IllegalArgumentException("GET failed,index need in [0,%d)".format(size))

    var cur = dummyHead.next
    for(i <- Range(0,index)){
        cur = cur.next
    }
    cur.e
  }

  def getFirst():T = {get(0)}
  def getLast():T = {get(size-1)}

  def set(index:Int,e:T): Unit ={
    if(index < 0 || index >= size) throw new IllegalArgumentException("SET failed,index need in [0,%d)".format(size))

    var cur = dummyHead.next
    for(i <- Range(0,index)){
        cur = cur.next
    }
    cur.e=e
  }

  def contains(e:T):Boolean = {
    var cur = dummyHead.next
    while (cur != null) {
      if (cur.e.equals(e)) {
        true
      }
      cur = cur.next
    }
    false
  }

  def remove(index:Int): T = {
    if(index < 0 || index >= size) throw new IllegalArgumentException("REmove failed,index need in [0,%d)".format(size))

    var prev = dummyHead
    for(i <- Range(0,index)){
        prev = prev.next
    }
    val retNode = prev.next
    prev.next = retNode.next
    retNode.next = null
    size-=1
    retNode.e
  }

  def removeFirst():T = {remove(0)}
  def removeLast():T = {remove(size-1)}

  def removeElement(e: T): Unit = {
    var prev = dummyHead
    var tp:Node = null
    while (prev.next != null && tp != null) {
      if (prev.next.e == e) tp = prev //todo: break is not supported
      prev = prev.next
    }
    if (prev.next != null) {
      val delNode = prev.next
      prev.next = delNode.next
      delNode.next = null
    }
  }

  override def toString: String = {
    var res = new StringBuilder()
    var cur = dummyHead.next
    while (cur != null) {
      res.append(cur + "->")
      cur = cur.next
    }
    res.append("NULL")
    res.toString()
  }

}

object LinkedList{
  def main(args: Array[String]): Unit = {

    var linkedList = new LinkedList[Int]()

    for(i <- Range(0,5)){
        linkedList.addFirst(i)
        println(linkedList)
    }
    linkedList.add(2,444)
    println(linkedList)

    linkedList.remove(2)
    println(linkedList)

    linkedList.removeFirst()
    println(linkedList)

    linkedList.removeLast()
    println(linkedList)


  }
}

File: src/main/java/Linked/LinkedList.java
package Linked;

public class  LinkedList<E> {

    private class Node{
        public E e;
        public Node next;

        public Node(E e,Node next){
            this.e=e;
            this.next = next;
        }

        public Node(E e){
            this(e,null);
        }
        public Node(){
            this(null,null);
        }

        @Override
        public String toString() {
            return e.toString();
        }
    }

    private Node dummyHead;
    private int size;

    public LinkedList(){
        dummyHead = new Node(null,null);
        size = 0;
    }

    public int getSize(){
        return size;
    }

    public boolean isEmpty(){
        return size==0;
    }
    public void add(int index,E e){
        if(index < 0 || index > size){
            throw new IllegalArgumentException(String.format("add failed,index need in [0,%d]", size));
        }

        Node prev = dummyHead;
        for (int i = 0; i < index; i++) {
            prev = prev.next;
        }
        prev.next = new Node(e,prev.next);
        size++;
    }
    public void addFirst(E e){
        add(0,e);
    }
    public void addLast(E e){
        add(size,e);
    }

    public E get(int index){
        if(index < 0 || index >= size){
            throw new IllegalArgumentException(String.format("Get faild, index need in [0,%d)", size));
        }
        Node cur = dummyHead.next;
        for (int i = 0; i < index; i++) {
            cur = cur.next;
        }
        return cur.e;
    }

    public E getFirst(){
        return get(0);
    }

    public E getLast(){
        return get(size-1);
    }

    public void set(int index,E e){
        if(index < 0 || index >= size){
            throw new IllegalArgumentException(String.format("Get faild, index need in [0,%d)", size));
        }
        Node cur = dummyHead.next;

        for (int i = 0; i < index; i++) {
            cur = cur.next;
        }
        cur.e = e;
    }

    public boolean contains(E e){
        Node cur = dummyHead.next;
        while (cur != null) {
            if (cur.e.equals(e)) {
                return true;
            }
            cur = cur.next;
        }
        return false;
    }

    public E remove(int index){
        if(index < 0 || index >= size){
            throw new IllegalArgumentException(String.format("Remove faild, index need in [0,%d)", size));
        }
        Node prev = dummyHead;
        for (int i = 0; i < index; i++) {
            prev = prev.next;
        }
        Node retNode = prev.next;
        prev.next=retNode.next;
        retNode.next=null;
        size--;
        return retNode.e;
    }

    public E removeFirst(){
        return remove(0);
    }

    public E removeLast(){
        return remove(size-1);
    }

    public void removeElement(E e){
        Node prev = dummyHead;
        while (prev.next != null){
            if(prev.next.e.equals(e)){break;}
            prev = prev.next;
        }

        if(prev.next != null){
            Node delNode = prev.next;
            prev.next = delNode.next;
            delNode.next = null;
        }

    }


    @Override
    public String toString() {
        StringBuilder res = new StringBuilder();

        for(Node cur = dummyHead.next ; cur != null ; cur = cur.next){
            res.append(cur + "->");
        }
        res.append("NULL");
        return res.toString();
    }

    public static void main(String[] args) {
        LinkedList<Integer> linkedList = new LinkedList<>();
        for (int i = 0; i < 5; i++) {
            linkedList.addFirst(i);
            System.out.println(linkedList);
        }
        linkedList.add(2,666);
        System.out.println(linkedList);

        linkedList.remove(2);
        System.out.println(linkedList);

        linkedList.removeFirst();
        System.out.println(linkedList);

        linkedList.removeLast();
        System.out.println(linkedList);

    }
}
